src/coordination/coordinator_data.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "coordination/coordinator_instance.hpp"
+#include "coordination/register_main_replica_coordinator_status.hpp"
+#ifdef MG_ENTERPRISE
+
+#include "coordination/coordinator_data.hpp"
+
+#include <range/v3/view.hpp>
+#include <shared_mutex>
+
+namespace memgraph::coordination {
+
+CoordinatorData::CoordinatorData() {
+  auto find_instance = [](CoordinatorData *coord_data, std::string_view instance_name) -> CoordinatorInstance & {
+    std::shared_lock<utils::RWLock> lock{coord_data->coord_data_lock_};
+
+    auto instance = std::ranges::find_if(
+        coord_data->registered_instances_,
+        [instance_name](const CoordinatorInstance &instance) { return instance.InstanceName() == instance_name; });
+
+    MG_ASSERT(instance != coord_data->registered_instances_.end(), "Instance {} not found during callback!",
+              instance_name);
+    return *instance;
+  };
+
+  replica_succ_cb_ = [find_instance](CoordinatorData *coord_data, std::string_view instance_name) -> void {
```

Comment:
This comment goes for all callbacks here.
1) Why do we create these callbacks? Can't we just put them inside the frequent check?
2) I don't think we want to `MG_ASSERT` here. We need to be supper robust. Check and handle it.

Reply:
If we really need to reuse the logic in these callbacks, why not create member functions?
Uses less space and is faster.

Reply:
Check comment about callbacks down

Reply:
> I don't think we want to MG_ASSERT here. We need to be supper robust. Check and handle it.

This situation should never ever happen. This is just defensive programming to assure I didn't make some big mistake. But instance executing replica callback must be an instance, no recovery here that's logic failure if it happens